### PR TITLE
 #114 Replace computeIfAbsent with custom logic

### PR DESCRIPTION
**Pull request by Voycawojka**
## Description
Removed `HashMap#computeIfAbset` that was causing the bug.

## Motivation and Context
https://github.com/Knotx/knotx-fragments/issues/114

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java [hunk 1]
```diff
@@ -65,15 +65,18 @@ public ActionProvider(Supplier<Iterator<ActionFactory>> supplier,
...
     } else {
       return Optional.of(createAction(action, actionFactoryOptions, factory));
     }
   }
 
-  private Function<String, Action> toAction(ActionFactoryOptions actionFactoryOptions,
-      ActionFactory factory) {
-    return action -> createAction(action, actionFactoryOptions, factory);
+  private Action cacheIfAbsent(String key, Action action) {
+    cache.putIfAbsent(key, action);
```
Comment (marcinus) [inline handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java]: Could you provide a reference that this modification does solve the concurrency issue?
I'm wondering that concurrent modification might still possible in the default HashMap implementation (didn't go through the source code though). How about using `ConcurrentHashMap`, or `Collections.synchronizedMap(new HashMap(...))` described in HashMap's JavaDoc?

Comment (Voycawojka) [reply handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java]: Using `ConcurrentHashMap` would only **kind of** solve the problem. The case is there should be no problem with concurrency here at all, but for some unknown reason there is. The `HashMap` implementation throws `ConcurrentModificationException` if it has been modified inside the function passed to `computeIfAbsent`. I therefore solved the problem by mimicking the same logic as with `computeIfAbsent` but in a non-atomic way. 

But we still don't know what exactly the core issue is, so it would probably be a good idea to find it out before merging.

Comment (Skejven) [reply handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java]: @Voycawojka according to #114 - which version of JDK do you use in the imported project in IntelliJ (I bet it is `> 8`, see why below).

It looks, that `ConcurrentModificationException ` was introduced since Java 9 to the[`HashMap.computeIfAbsent()`](https://docs.oracle.com/javase/9/docs/api/java/util/HashMap.html#computeIfAbsent-K-java.util.function.Function-) method (or at least it is not mentioned in Java 8 docs).
It says:
> This method will, on a best-effort basis, throw a ConcurrentModificationException if it is detected that the mapping function modifies this map during computation.

And the exception is thrown because of recurrence (`createAction()` is invoking `get()` method, which can again all `createAction()`).
I've dug in a little and it looks, that there was slight change of the behaviour in `computeIfAbsent` between Java 8 and 9, because of this bug: https://bugs.openjdk.java.net/browse/JDK-8062841 (looks like some kind of guard to not end with endless loop).

Long story short -> it looks that your implementation is the proper way to address the recursive operation modifying the `cache` (HashMap).

File: handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java [hunk 2]
```diff
@@ -65,15 +65,18 @@ public ActionProvider(Supplier<Iterator<ActionFactory>> supplier,
     }
 
     if (isCacheable(factory)) {
-      return Optional.of(cache.computeIfAbsent(action, toAction(actionFactoryOptions, factory)));
+      return Optional.ofNullable(cache.get(action))
+          .map(Optional::of)
+          .orElseGet(() -> Optional.ofNullable(createAction(action, actionFactoryOptions, factory)))
+          .map(createdAction -> cacheIfAbsent(action, createdAction));
```
Comment (Skejven) [inline handler/core/src/main/java/io/knotx/fragments/task/factory/node/action/ActionProvider.java]: Why not just:
```suggestion
          .map(createdAction -> cache.putIfAbsent(action, createdAction));
```
?
